[mlir][inliner] Add doClone and canHandleMultipleBlocks callbacks to Inliner

junfengd-nv opened the pull request:
Current inliner disables inlining when the caller is in a region with single block trait, while the callee function contains multiple blocks. the SingleBlock trait is used in operations such as do/while loop, for example fir.do_loop, fir.iterate_while and fir.if (refer to https://flang.llvm.org/docs/FIRLangRef.html). Typically, calls within loops are good candidates for inlining. However, functions with multiple blocks are also common. for example, any function with "if () then return" will result in multiple blocks in MLIR.

My change gives the flexibility of a customized inliner to handle such cases. 
doClone: clones instructions and other information from the callee function into the caller function. . 
canHandleMultipleBlocks: checks if functions with multiple blocks can be inlined into a region with the SingleBlock trait.
With the two callback functions that I added, the derived inliner can return true with function "canHandleMultipleBlocks" to avoid stopping during the inline decision process. And then add their extra transformations in the function "doClone".

The default behavior of the inliner remains unchanged.

junfengd-nv committed:
[mlir][inliner] Add doClone and canHandleMultipleBlocks callbacks to the Inliner

These callback functions enhance the flexibility to customize inliner behavior.
. doClone: clones instructions and other information from the callee function into the caller function.
. canHandleMultipleBlocks: checks if functions with multiple blocks can be inlined into a region with the SingleBlock trait.

The default behavior of the inliner remains unchanged.

jeanPerier commented on mlir/lib/Transforms/InlinerPass.cpp:
```diff
@@ -142,9 +142,13 @@ void InlinerPass::runOnOperation() {
     return isProfitableToInline(call, inliningThreshold);
   };
 
+  // By default, prevent inlining a functon containing multiple blocks into a
```
```suggestion
  // By default, prevent inlining a function containing multiple blocks into a
```

jeanPerier reviewed (commented):
Makes sense to me. Letting people more familiar with this code reviewing/approving.

It would be great to be able to add a test showing the motivation of this interface by having a Clone version that can generate an scf.execute_region, but I understand it is not desired to make the InlinerPass dependent on the SCF dialect for the sake of this test.

joker-eph commented:
I don't quite get the motivation for these and the description isn't clear on this: can you elaborate on why are these useful / desirable?

junfengd-nv committed:
Update mlir/lib/Transforms/InlinerPass.cpp

Co-authored-by: jeanPerier <jean.perier.polytechnique@gmail.com>

joker-eph commented:
Can you please:
1) Add this information to the PR description, this is very clear, thanks.
2) Add tests for this in the test dialect?

junfengd-nv commented:
> 1. Add this information to the PR description, this is very clear, thanks.
Updated
> 2. Add tests for this in the test dialect?
The default behavior of MLIR inliner doesn't change, so I have not changed and added any tests. Do you want an example of demonstrating the creation of a customized inliner with wrapping in scf.execute_region? or just tests showing  doClone/canHandleMultipleBlocks could be overridden? could you please pointing me an existing example if it exists? Thanks

joker-eph commented:
> . Do you want an example of demonstrating the creation of a customized inliner with wrapping in scf.execute_region?

I think that would be a good test to have!

junfengd-nv committed:
Add a test for Inliner callback functions.

junfengd-nv commented:
@joker-eph @jeanPerier I have added a test under the test dialect to demonstrate the concept of customizing the inliner to wrap the callee into an SCF execute region. The SCF dialect has been added as a dependency solely for that test pass, and not for the default inliner pass.

Could you please review the changes again? Thank you very much.

joker-eph commented on mlir/lib/Transforms/Utils/Inliner.cpp:
```diff
@@ -344,6 +344,28 @@ static void collectCallOps(iterator_range<Region::iterator> blocks,
+//===----------------------------------------------------------------------===//
+// Initialize doClone function with the default implementation
+Inliner::CloneCallbackTy &Inliner::doClone() {
+  static Inliner::CloneCallbackTy doWork =
```
This can't work: you're using basically a single global variable that will be used for the entire process, shared by all inliner instance. That would break many use-cases (and isn't thread-safe).

joker-eph commented on mlir/include/mlir/Transforms/Inliner.h:
```diff
@@ -90,18 +90,40 @@ class Inliner {
+  /// default, it is not allowed.
+  /// If this function return true, the static member function doClone()
+  /// should perform the actual transformation with its support.
+  using canHandleMultipleBlocksCbTy = std::function<bool()>;
```
Should we just move this (and the doClone) to the `InlinerConfig` instead?
Let the user override the default callback in the config as needed.

(see below for the design issue with what you have right now).

junfengd-nv replied:
That is a good suggestion. I will move them.

junfengd-nv committed:
Move to InlinerConfig

junfengd-nv committed:
fix format.

junfengd-nv committed:
Remove unused codes

joker-eph commented on mlir/include/mlir/Transforms/Inliner.h:
```diff
@@ -60,6 +73,28 @@ class InlinerConfig {
+                                       src->getBlocks(), src->begin(),
+                                       src->end());
+  };
+  /// Determining if the inliner can inline a function containing multiple
```
```suggestion
  /// Determine if the inliner can inline a function containing multiple
```

joker-eph commented on mlir/include/mlir/Transforms/Inliner.h:
```diff
@@ -60,6 +73,28 @@ class InlinerConfig {
+  };
+  /// Determining if the inliner can inline a function containing multiple
+  /// blocks into a region that requires a single block. By default, it is
+  /// not allowed. If it is true, cloneCallback shuold perform the extra
```
```suggestion
  /// not allowed. If it is true, cloneCallback should perform the extra
```

joker-eph commented on mlir/include/mlir/Transforms/InliningUtils.h:
```diff
@@ -253,13 +254,15 @@ class InlinerInterface
 /// region should be cloned into the 'inlinePoint' or spliced directly.
-LogicalResult inlineRegion(InlinerInterface &interface, Region *src,
+LogicalResult inlineRegion(InlinerInterface &interface,
+                           const InlinerConfig &config, Region *src,
```
Requiring the entire config seems overkill to me. We should be able to just take a `function_ref` that has the signature of the callback here.

joker-eph commented on mlir/include/mlir/Transforms/Inliner.h:
```diff
@@ -39,13 +43,22 @@ class InlinerConfig {
+  void setCloneCallback(CloneCallbackTy callback) {
+    cloneCallback = std::move(callback);
+  }
+  void setCanHandleMultipleBlocks(bool value) {
```
```suggestion
  void setCanHandleMultipleBlocks(bool value = true) {
```

joker-eph commented on mlir/test/lib/Transforms/TestInliningCallback.cpp:
```diff
@@ -0,0 +1,151 @@
+                    runPipelineHelper, config, profitabilityCb);
+
+    // Collect each of the direct function calls within the module.
+    SmallVector<func::CallIndirectOp, 16> callers;
```
```suggestion
    SmallVector<func::CallIndirectOp> callers;
```

joker-eph commented on mlir/test/lib/Transforms/TestInliningCallback.cpp:
```diff
@@ -0,0 +1,151 @@
+    InlinerConfig config;
+    CallGraph &cg = getAnalysis<CallGraph>();
+
+    auto function = getOperation();
```
```suggestion
    func::FuncOp function = getOperation();
```

junfengd-nv committed:
Update mlir/include/mlir/Transforms/Inliner.h

Co-authored-by: Mehdi Amini <joker.eph@gmail.com>

junfengd-nv committed:
Update mlir/include/mlir/Transforms/Inliner.h

Co-authored-by: Mehdi Amini <joker.eph@gmail.com>

junfengd-nv committed:
Update mlir/include/mlir/Transforms/Inliner.h

Co-authored-by: Mehdi Amini <joker.eph@gmail.com>

junfengd-nv committed:
Update mlir/test/lib/Transforms/TestInliningCallback.cpp

Co-authored-by: Mehdi Amini <joker.eph@gmail.com>

junfengd-nv committed:
Update mlir/test/lib/Transforms/TestInliningCallback.cpp

Co-authored-by: Mehdi Amini <joker.eph@gmail.com>

junfengd-nv committed:
Update mlir/test/lib/Transforms/TestInliningCallback.cpp

Co-authored-by: Mehdi Amini <joker.eph@gmail.com>

junfengd-nv committed:
Pass cloneCallback function referece instead of whole InlinerConfig class

joker-eph commented on mlir/include/mlir/Transforms/InliningUtils.h:
```diff
@@ -255,16 +255,17 @@ class InlinerInterface
-                           const InlinerConfig &config, Region *src,
-                           Operation *inlinePoint, IRMapping &mapper,
-                           ValueRange resultsToReplace,
+                           InlinerConfig::CloneCallbackTy cloneCallback,
```
Unfortunately this isn't a function_ref, and copying a std::function around isn't just "free".
I'd rather see a function_ref here.

junfengd-nv replied:
Thanks for pointing this out, I updated the change by using "function_ref<InlinerConfig::CloneCallbackSigTy> cloneCallback"  instead.

joker-eph commented on mlir/include/mlir/Transforms/Inliner.h:
```diff
@@ -27,6 +27,10 @@ class InlinerConfig {
 public:
   using DefaultPipelineTy = std::function<void(OpPassManager &)>;
   using OpPipelinesTy = llvm::StringMap<OpPassManager>;
+  using CloneCallbackTy =
+      std::function<void(OpBuilder &builder, Region *src, Block *inlineBlock,
+                         Block *postInsertBlock, IRMapping &mapper,
+                         bool shouldCloneInlinedRegion)>;
```
```suggestion
  using CloneCallbackSigTy =
      void(*)(OpBuilder &builder, Region *src, Block *inlineBlock,
                         Block *postInsertBlock, IRMapping &mapper,
                         bool shouldCloneInlinedRegion);
  using CloneCallbackTy =
      std::function<CloneCallbackSigTy>;
```

Splitting it like this will allow to easily declare the inlineRegion function as taking a `function_ref< InlinerConfig::CloneCallbackSigTy>`.

junfengd-nv committed:
Update mlir/include/mlir/Transforms/Inliner.h

Co-authored-by: Mehdi Amini <joker.eph@gmail.com>

junfengd-nv committed:
Correct the parameter to use function_ref.

joker-eph reviewed (approved):
LGTM, thanks!

joker-eph commented on mlir/include/mlir/Transforms/InliningUtils.h:
```diff
@@ -253,43 +254,51 @@ class InlinerInterface
 /// provided, will be used to update the inlined operations' location
 /// information. 'shouldCloneInlinedRegion' corresponds to whether the source
 /// region should be cloned into the 'inlinePoint' or spliced directly.
```
Actually, the documentation the all the API changes should be updated, can you do it in a follow-up PR please?

junfengd-nv replied:
Yes. I will.
